### ISSUE TITLE
weight pruning at high sparsity

### ISSUE THREAD

**Issue by chrishamblin7** [chrishamblin7/circuit_pruner#1]
When extracting a circuit with 'weight' structure at high sparsity returned circuits sometimes have filters that are fully masked but not removed . . .

```python
#hardware
device = 'cuda:0'

from circuit_pruner.utils import load_config

config = load_config('/mnt/data/chris/dropbox/Research-Hamblin/Projects/circuit_pruner_cvpr2022/configs/alexnet_sparse_config.py')
model = config.model.to(device)



#general
import torch
import os
from copy import deepcopy
from circuit_pruner.force import setup_net_for_circuit_prune, show_model_layer_names


show_model_layer_names(model)

#feature target
# layer = 'features_10'
# unit = 241
layer = 'features_6'
unit = 222
#layer = 'features_0'
#unit = 0

feature_targets = {layer:[unit]}


from PIL import Image, ImageOps

image_folder = '../../image_data/sample_data/mask_featviz/features_6_222_gencircuit/'
image_files = os.listdir(image_folder)


for file in image_files:
    print(file)
    image = Image.open(image_folder+file)
    image = image.resize((224,224))
    display(image)

    
    
from circuit_pruner.data_loading import single_image_data
from torchvision import transforms
from torch.utils import data


preprocess =  transforms.Compose([
                                transforms.Resize((224,224)),
                                transforms.ToTensor(),
                                transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                                     std=[0.229, 0.224, 0.225])])
# preprocess =  transforms.Compose([
#                                 transforms.Resize((224,224)),
#                                 transforms.ToTensor()])

#make single image data loaders
kwargs = {'num_workers': 1, 'pin_memory': True, 'sampler':None} if 'cuda' in device else {}


data_loaders = {}
for file in image_files:
    data_loaders[file] = data.DataLoader(single_image_data(image_folder+file,preprocess),
                                                batch_size=1,
                                                shuffle=False,
                                                **kwargs)
                  
        


from circuit_pruner.force import circuit_FORCE_pruning

masks = {}
ranks = {}


file = 't1.jpg'

structure='weights'

data_loader = data_loaders[file]


weight_rank, weight_mask = circuit_FORCE_pruning(model, data_loader, feature_targets = feature_targets, 
                                   T=10, keep_ratio=.1, rank_field='max',
                                   use_abs_ranks=True, return_ranks=True, structure=structure)


for i in range(len(weight_rank)):
    weight_rank[i] = weight_rank[i].to('cpu')
    weight_mask[i] = weight_mask[i].to('cpu')  

torch.cuda.empty_cache()

    


#finally lets extract some circuits
#from circuit_pruner.extraction import dissected_model_ranks_2_circuit_model


img = Image.open(image_folder+file)
img= img.resize((224,224))
display(img)

sparsity = .001


circuit, mask = dissected_model_ranks_2_circuit_model(weight_rank,sparsity,
                                                            model,feature_targets,
                                                            device,structure='weights')

circuit.features_0.weight[4]

```
